Framework/Parallel/test/EventParserTest.h

```diff
@@ -0,0 +1,464 @@
+    gen.cleanupEventLists(eventLists);
+  }
+
+  void testParsingFull_InParts_1Rank_3Banks() {
+    size_t numBanks = 3;
+    detail::FakeParserDataGenerator<int32_t, int64_t, double> gen(3, 20, 7);
+    auto eventLists = gen.prepareEventLists(gen.eventLists().size());
+    auto parser = gen.createSimpleParser(eventLists, gen.bankOffsets());
+
+    for (size_t bank = 0; bank < numBanks; bank++) {
+      parser.setPulseInformation(gen.eventIndex(bank), gen.eventTimeZero());
+      auto event_id = gen.eventId(bank);
+      auto event_time_offset = gen.eventTimeOffset(bank);
+
+      auto parts = 11;
+      auto portion = event_id.size() / parts;
+
+      for (int i = 0; i < parts; ++i) {
+        auto offset = portion * i;
+
+        // Needed so that no data is missed.
+        if (i == (parts - 1))
+          portion = event_id.size() - offset;
+
+        LoadRange range{bank, offset, portion};
+        parser.startParsing(event_id.data() + offset,
+                            event_time_offset.data() + offset, range);
+        parser.wait();
+      }
+    }
+    parser.finalize();
+    compareEventLists(gen.eventLists(), eventLists);
+    gen.cleanupEventLists(eventLists);
+  }
+
+private:
+  void compareEventLists(const std::vector<std::vector<TofEvent> *> &lhs,
+                         const std::vector<std::vector<TofEvent> *> &rhs) {
+    for (size_t i = 0; i < lhs.size(); ++i)
+      TS_ASSERT_EQUALS(*rhs[i], *lhs[i]);
+  }
+
+  template <typename IndexType, typename TimeZeroType, typename TimeOffsetType>
+  void
+  doTestRankData(const std::vector<std::vector<Event>> &rankData,
+                 EventParser<IndexType, TimeZeroType, TimeOffsetType> &parser,
+                 detail::FakeParserDataGenerator<IndexType, TimeZeroType,
+                                                 TimeOffsetType> &gen,
+                 const LoadRange &range) {
+    size_t cur = 0;
+    auto res = parser.findStartAndEndPulseIndices(
+        gen.eventIndex(0), range.eventOffset, range.eventCount, cur);
+
+    for (size_t pulse = res.first; pulse <= res.second; ++pulse) {
+      auto start =
+          std::max(pulse == 0
+                       ? 0
+                       : static_cast<size_t>(gen.eventIndex(0)[pulse - 1]),
+                   range.eventOffset) -
+          range.eventOffset;
+      auto end = std::min(static_cast<size_t>(gen.eventIndex(0)[pulse] - 1),
+                          range.eventOffset + range.eventCount) -
+                 range.eventOffset;
+      auto &pulses = gen.eventTimeZero();
+      TS_ASSERT(std::all_of(rankData[0].cbegin() + start,
+                            rankData[0].cbegin() + end,
+                            [pulses, pulse](const Event &e) {
+                              return e.tofEvent.pulseTime() ==
+                                     static_cast<int64_t>(pulses[pulse]);
+                            }));
+    }
+  }
+};
+
+// TODO Add performance Tests
```

Comment:
Are you going to add performance tests as part of this PR?

Reply:
Yes, currently writing those.